### ISSUE TITLE
Rule proposal: no-union-in-omit Disallow omit for union types

### ISSUE THREAD

**Issue by kecrily** [typescript-eslint/typescript-eslint#7974]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

`Omit` is forbidden for `Union` type because the result will always be `{}`

### Fail Cases

```typescript
type User = {
  id: string
  name: string
}

type Product = {
  id: string
  price: number
}

type EntityWithout = Omit<User | Product, "id">
```

### Pass Cases

```typescript
type User = {
  id: string
  name: string
}

type EntityWithout = Omit<User, "id">
```

### Additional Info

Inspried by https://twitter.com/mattpocockuk/status/1727262093400023095

- [X] I am willing to submit a pull request to implement this change.

**Comment by bradzacher** [comment]
> Omit is forbidden for Union type because the result will always be {}

This is not correct.

The way that `Omit` works is that it filters `keyof T`. When `T` is a union type then TS returns the intersection of the keys - i.e. TS returns the keys that exist in all types in the union.

So in your example the only common key is `id` - so if you Omit `id` you are left with no keys and thus are left with the empty object.

I don't know if a lint rule makes sense here because sometimes you do want this behaviour and sometimes you don't. If you don't then the distributive form that Matt described later in the thread will do the job you want instead.

**Comment by JoshuaKGoldberg** [comment]
It's been ~4 months since last comment and there haven't been any new motivators for this. Closing it out as having not had enough community feedback to suggest we should take action.

If anybody sees this and thinks they have a good reason not yet mentioned, please do file a new issue! We'd be happy to take a look.

Thanks for the discussion all!
